PP-1328: @requirements data in PTL json test report

shilpakodli opened the pull request:
#### Bug/feature Description
* *[PP-1328](https://pbspro.atlassian.net/browse/PP-1328)*

* Data relating to Test case requirements specified in @requirements decorator needs to be included in PTL's json test report's "requirements" field.
* Currently it lists a blank dictionary as below 
{
"testsuites": {
"TestRequirementsDecorator": {
"testcases": {
"test_tc_run": {
"docstring": "Test to verify test run when requirements are satisfied test suite requirements overridden by test case requirements",
"requirements": {},
"results":

{ "status": "PASS", "start_time": "2019-02-11 00:24:15.497798", "measurements": [], "status_data": "", "end_time": "2019-02-11 00:24:17.992014", "duration": "0:00:02.494216" }


#### Affected Platform(s)
* All

#### Solution Description
*  Include data related to @requirements decorator at test suite and test case level into json report.

#### Testing logs/output
* [ptl_execution_logs.txt](https://github.com/PBSPro/pbspro/files/2956477/ptl_execution_logs.txt)
* [ptl_test_results_json.txt](https://github.com/PBSPro/pbspro/files/2959937/ptl_test_results_json.txt)



#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

saritakh commented on test/tests/selftest/pbs_requirements_decorator.py:
```diff
@@ -89,3 +89,19 @@ def test_skip_no_server_on_mom(self):
         """
         self.server.expect(SERVER, {'server_state': 'Active'})
+
+    @requirements(num_servers=1, num_comms=1, no_comm_on_mom=True)
```
'no_comm_on_mom' is True by default, this can be changed to False. Although this is not testing the change in json fields, it is an extended test for requirements decorator. I think test case -TestJSONReport.test_json_fields (pbs_json_test_report.py) can be extended to verify requirements field; which would validate the data populated in PTLJsonData.

shilpakodli commented:
@saritakh I have addressed your review comments and attaching new test execution logs.
[ptl_json_report_test_logs.txt](https://github.com/PBSPro/pbspro/files/2965937/ptl_json_report_test_logs.txt)

saritakh commented on test/fw/ptl/utils/plugins/ptl_report_json.py:
```diff
@@ -118,7 +119,8 @@ def get_json(self, data, prev_data=None):
             'measurements': []
         }
-        tcshort['requirements'] = data['requirements']
+        tcshort['requirements'] = getattr(data, 'requirements',
```
We shouldn't be doing this. PTLJsonData.get_json() is just a converter of test data passed to it in dictionary format into json format. It should not be creating or populating data in this method.

saritakh commented on test/tests/selftest/pbs_json_test_report.py:
```diff
@@ -118,6 +127,10 @@ def test_json_fields(self):
             for t in jdata['testsuites'][s]:
                 for u in jdata['testsuites'][s]['testcases']:
                     for r in verify_data['test_cases_info']:
+                        if r == 'requirements':
```
This loop previously checked only the keys of test case information. I suggest to move this field value check to 'for' loop in line 142 "for k, v in field_values.items():"; since that is the place where the test checks for other field values - machine_info and test result status.

shilpakodli commented:
@saritakh  Thanks Sarita. I have made changes and pushed my changes. Please review it.

saritakh commented:
@shilpakodli Line 40 - import of default_requirements can be removed from pbs_report_json.py. Also, please post the ptl test logs of the test case.

shilpakodli commented:
@saritakh Attaching test execution logs and also addressed review comments.
[ptl_json_report_test_logs_1.txt](https://github.com/PBSPro/pbspro/files/2969741/ptl_json_report_test_logs_1.txt)
